_posts/2020-05-27-mlr3keras-boston-housing-task/mlr3keras-boston-housing-task.Rmd

```diff
@@ -0,0 +1,332 @@
+      layer_dense(units = 1L)
+  
+    model %>% compile(
+    loss = "mse",
+    optimizer = optimizer_rmsprop(),
+    metrics = list("mean_absolute_error")
+  )
+
+  model
+}
+
+model = build_model()
+model %>% summary()
+```
+
+For this use case we create a sequential model, a linear stack of three layers.
+The shape of our input data is defined in the input layer.
+Input is analyzed by two densely connected hidden layers.
+A final denselayer returns a single continuous target estimation value.
+
+# Learner
+
+### Define hyperparameters
+
+Heading back to `r mlr_pkg("mlr3")` syntax  we set up the keras regression learner like any other and define hyperparameters.
+Note that the compiled `model` defined in the previous chapter needs to be defined as a hyperparameter in the learners' `param_set`.
+
+```{r hyperparams}
+learner = po("learner", lrn("regr.keras"))
+learner = lrn("regr.keras")
+
+learner$param_set$values$epochs = 500L
+learner$param_set$values$model = build_model()
+learner$param_set$values$validation_split = 0.2
+```
+
+### Callbacks
+
+Another hypereparameter is `callback`.
+A `callback` is a set of functions, bundled in a list that can be used to explore specified progress stages during training.
+Accordingly a defined `callback` function is called at its' specified stage of training to monitor or display specified model parameters at this stage.
+The history callback for example is applied to each training progress of a model by default.
+It stores the training history in a `history` slot of its' concerning model object.
+For a list of all implemented callbacks see [Keras Callbacks](https://keras.io/callbacks/).
+
+In our first learner we want to reduce information about training progress to a single dot printed out at the end of each applied epoch.
+Therefore we create a `LambdaCallback` that is called at each epochs end and assign it to the learners' `callbacks`-hyperparameter.
+
+```{r callbacks}
+print_dot_callback <- callback_lambda(
+  on_epoch_end = function(epoch, logs) {
+    if (epoch %% 80 == 0) cat("\n")
+    cat(".")
+  }
+) 
+
+learner$param_set$values$callbacks = list(print_dot_callback)
+```
+
+Callback can also be called through `r gh_pkg("mlr-org/mlr3keras")` syntax.
+As not all callbacks are implemented yet [mlr3keras callbacks](https://github.com/mlr-org/mlr3keras/blob/master/R/keras_callbacks.R) is still in progress.
+Currently working callbacks within `r gh_pkg("mlr-org/mlr3keras")`-syntax  are:
+
+* `cb_es`: Early stopping callback
+* `cb_lrs`: Learning rate scheduler callback
+* `cb_tb`: Tensorboard callback
+* `cb_lr_log`: Learning rate logger callback
+* `LogMetrics`: Batch-wise Metrics Logger Callback
+* `SetLogLR`: Batch-wise Metrics Setter Callback
+
+Initialize early stopping like this:
+
+```{r mlr3cb}
+# early_stop <- callback_early_stopping(monitor = "val_loss", patience = 20) # keras syntax
+early_stop <- cb_es(monitor = "val_loss", patience = 20) # mlr3 syntax
```

Comment:
We mix up `<-` und `=` throughout the article, please choose one and be consistent